src/index/VocabularyMerger.h

```diff
@@ -16,12 +17,106 @@
 #include "index/IndexBuilderTypes.h"
 #include "index/Vocabulary.h"
 #include "util/HashMap.h"
-#include "util/MmapVector.h"
 #include "util/ProgressBar.h"
+#include "util/Serializer/FileSerializer.h"
+#include "util/Serializer/SerializePair.h"
+#include "util/Serializer/SerializeVector.h"
 #include "util/TypeTraits.h"
 
-using IdPairMMapVec = ad_utility::MmapVector<std::pair<Id, Id>>;
-using IdPairMMapVecView = ad_utility::MmapVectorView<std::pair<Id, Id>>;
+// Writes pairs of (partial ID, global ID) incrementally to a file
+class IdMap {
+ private:
+  uint64_t count_ = 0;
+  std::string filename_;
+  std::unique_ptr<ad_utility::serialization::FileWriteSerializer> serializer_;
+  bool headerWritten_ = false;
+
+ public:
+  explicit IdMap(const std::string& filename) : filename_(filename) {
+    serializer_ =
+        std::make_unique<ad_utility::serialization::FileWriteSerializer>(
+            filename);
+  }
+
+  IdMap(size_t, const std::string& filename) : IdMap(filename) {}
+
+  void push_back(const std::pair<Id, Id>& pair) {
+    // Write header (vector size) on first push_back - we'll update it in
+    // destructor
+    if (!headerWritten_) {
+      // Reserve space for size, will be updated in destructor
+      uint64_t placeholder = 0;
+      *serializer_ << placeholder;
+      headerWritten_ = true;
+    }
+    // Write the pair directly to file
+    *serializer_ << pair;
+    ++count_;
+  }
+
+  ~IdMap() {
+    if (serializer_) {
+      try {
+        // Update the size at the beginning of the file
+        auto currentPos = serializer_->getSerializationPosition();
+        serializer_->setSerializationPosition(0);
+        *serializer_ << count_;
+        serializer_->setSerializationPosition(currentPos);
+      } catch (...) {
+        // Handle errors gracefully in destructor
+      }
+    }
+  }
+
+  // Disable copy constructor and assignment
+  IdMap(const IdMap&) = delete;
+  IdMap& operator=(const IdMap&) = delete;
+
+  // Allow move operations
+  IdMap(IdMap&& other) noexcept
+      : count_(other.count_),
+        filename_(std::move(other.filename_)),
+        serializer_(std::move(other.serializer_)),
+        headerWritten_(other.headerWritten_) {
+    other.headerWritten_ = false;
+    other.count_ = 0;
+  }
+
+  IdMap& operator=(IdMap&& other) noexcept {
+    if (this != &other) {
+      count_ = other.count_;
+      filename_ = std::move(other.filename_);
+      serializer_ = std::move(other.serializer_);
+      headerWritten_ = other.headerWritten_;
+      other.headerWritten_ = false;
+      other.count_ = 0;
+    }
+    return *this;
+  }
+};
+
+// Read-only view of pairs of (partial ID, global ID) deserialized from a file
+class IdMapView {
+ private:
+  std::vector<std::pair<Id, Id>> data_;
+
+ public:
+  explicit IdMapView(const std::string& filename) {
+    ad_utility::serialization::FileReadSerializer serializer(filename);
+    serializer >> data_;
```

Comment:
Since this is the core functionality of this class, just make this a free function and return a `std::vector<std::pair<Id, Id>>`. Then this class is not needed at all. That is of course if it's on purpose that everything is loaded into memory at once.